extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -533,4 +464,65 @@ private boolean comparePartitionKeys(ResultRow row1, ResultRow row2, List<String
     }
     return match == partitionColumnNames.size();
   }
+
+  private void makeRowSupplierFromFrameCursor()
+  {
+    final ColumnSelectorFactory frameColumnSelectorFactory = frameCursor.getColumnSelectorFactory();
+    final Supplier<Object>[] fieldSuppliers = new Supplier[frameReader.signature().size()];
+    for (int i = 0; i < fieldSuppliers.length; i++) {
+      final ColumnValueSelector<?> selector =
+          frameColumnSelectorFactory.makeColumnValueSelector(frameReader.signature().getColumnName(i));
+      fieldSuppliers[i] = selector::getObject;
+    }
+    rowSupplierFromFrameCursor = () -> {
+      final ResultRow row = ResultRow.create(fieldSuppliers.length);
+      for (int i = 0; i < fieldSuppliers.length; i++) {
+        row.set(i, fieldSuppliers[i].get());
+      }
+      return row;
+    };
+  }
+
+  /**
+   * This method is responsible for the logic of batching the PARTITION BY keys.
+   * If we can add the current rows (currentBatchOfRacs) to the rows pending processing (batchOfRacsPendingProcessing) without violating the maxRowsMaterialized constraint, do it.
+   * If we cannot, we process the rows pending processing (batchOfRacsPendingProcessing), and add the current rows to it which would then be processed in future.
+   */
+  private void processBatchOfRacsPendingProcessingIfNeeded()
+  {
+    if (currentBatchOfRacs.size() + batchOfRacsPendingProcessing.size() > maxRowsMaterialized) {
+      processBatchOfRacsPendingProcessing();
+    }
+
+    batchOfRacsPendingProcessing.addAll(currentBatchOfRacs);
```

Comment:
This is a fairly expensive call. 
can we avoid this?

Reply:
I thought about this. Having an approach where we use a single ArrayList, and maintain the last PARTITION BY boundary index also has the same issue, as we would need to collect all rows until that index into a new ArrayList before sending it to the list of operators.

> is there an approach where we can avoid this. Swapping buffers maybe

Can you elaborate more on swapping buffers approach? We would still have to copy the contents of the buffer into a new ArrayList before sending it to the list of operators (hence degenerating into the `last PARTITION BY boundary index` alternative approach I described above, IIUC)

Also, since we anyway have the next step to move to a GlueingPartitioningOperator approach, I feel we shouldn't invest too much time into designing and over-engineering this patch, and keep things simple for the time being. (as GlueingPartitioningOperator approach would get rid of this logic entirely from this layer, and would solve this concern) - Thoughts?

Reply:
On further thoughts, we can change the approach to use a single Arraylist, and maintain the last PARTITION BY boundary index, and add a new method `MapOfColumnsRowsAndColumns.fromResultRowTillIndex()` so that we don't have to separately create the sublist to pass to `MapOfColumnsRowsAndColumns.fromResultRow()`.

I ended up making the changes locally for this, the diff looks like the following:

```diff
diff --git a/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java b/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java
index 9945274bc5..54018893cc 100644
--- a/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java
+++ b/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java
@@ -94,13 +94,8 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
   // Type strategies are pushed in the same order as column types in frameReader.signature()
   private final NullableTypeStrategy[] typeStrategies;
 
-  // currentBatchOfRacs holds the current rows being read from the frame.
-  private final ArrayList<ResultRow> currentBatchOfRacs;
-
-  // batchOfRacsPendingProcessing holds the rows read from the frame that are pending processing.
-  // These get processed when we either reach the end of input channel, or when we meet the criteria in shouldProcessPendingBatch() method.
-  // Until then, this accumulates rows from currentBatchOfRacs.
-  private final ArrayList<ResultRow> batchOfRacsPendingProcessing;
+  private final ArrayList<ResultRow> rowsToProcess;
+  private int lastPartitionIndex = 0;
 
   public WindowOperatorQueryFrameProcessor(
       WindowOperatorQuery query,
@@ -123,8 +118,7 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
     this.query = query;
     this.frameRowsAndCols = new ArrayList<>();
     this.resultRowAndCols = new ArrayList<>();
-    this.currentBatchOfRacs = new ArrayList<>();
-    this.batchOfRacsPendingProcessing = new ArrayList<>();
+    this.rowsToProcess = new ArrayList<>();
     this.maxRowsMaterialized = maxRowsMaterializedInWindow;
     this.partitionColumnNames = partitionColumnNames;
 
@@ -244,8 +238,8 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
         makeRowSupplierFromFrameCursor();
       } else if (inputChannel.isFinished()) {
         // Handle any remaining data.
-        batchOfRacsPendingProcessing.addAll(currentBatchOfRacs);
-        processBatchOfRacsPendingProcessing();
+        lastPartitionIndex = rowsToProcess.size();
+        processRowsUpToLastPartition();
         return ReturnOrAwait.returnObject(Unit.instance());
       } else {
         return ReturnOrAwait.runAgain();
@@ -256,20 +250,16 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
       final ResultRow currentRow = rowSupplierFromFrameCursor.get();
       if (outputRow == null) {
         outputRow = currentRow;
-        currentBatchOfRacs.add(currentRow);
+        rowsToProcess.add(currentRow);
       } else if (comparePartitionKeys(outputRow, currentRow, partitionColumnNames)) {
         // Add current row to the same batch of rows for processing.
-        currentBatchOfRacs.add(currentRow);
-        ensureMaxRowsInAWindowConstraint(currentBatchOfRacs.size());
-        if (shouldProcessPendingBatch()) {
+        rowsToProcess.add(currentRow);
+        if (rowsToProcess.size() > maxRowsMaterialized) {
           // We don't want to materialize more than maxRowsMaterialized rows at any point in time, so process the pending batch.
-          processBatchOfRacsPendingProcessing();
+          processRowsUpToLastPartition();
         }
       } else {
-        if (shouldProcessPendingBatch()) {
-          processBatchOfRacsPendingProcessing();
-        }
-        copyAndClearCurrentBatch();
+        lastPartitionIndex = rowsToProcess.size();
         outputRow = currentRow.copy();
         return ReturnOrAwait.runAgain();
       }
@@ -486,35 +476,6 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
     };
   }
 
-  /**
-   * Process {@link #batchOfRacsPendingProcessing}.
-   */
-  private void processBatchOfRacsPendingProcessing()
-  {
-    if (batchOfRacsPendingProcessing.isEmpty()) {
-      return;
-    }
-    RowsAndColumns singleRac = MapOfColumnsRowsAndColumns.fromResultRow(
-        batchOfRacsPendingProcessing,
-        frameReader.signature()
-    );
-    ArrayList<RowsAndColumns> rowsAndColumns = new ArrayList<>();
-    rowsAndColumns.add(singleRac);
-    runAllOpsOnMultipleRac(rowsAndColumns);
-    batchOfRacsPendingProcessing.clear();
-  }
-
-  private boolean shouldProcessPendingBatch()
-  {
-    return currentBatchOfRacs.size() + batchOfRacsPendingProcessing.size() > maxRowsMaterialized;
-  }
-
-  private void copyAndClearCurrentBatch()
-  {
-    batchOfRacsPendingProcessing.addAll(currentBatchOfRacs);
-    currentBatchOfRacs.clear();
-  }
-
   private void ensureMaxRowsInAWindowConstraint(int numRowsInWindow)
   {
     if (numRowsInWindow > maxRowsMaterialized) {
@@ -524,4 +485,22 @@ public class WindowOperatorQueryFrameProcessor implements FrameProcessor<Object>
       ));
     }
   }
+
+  private void processRowsUpToLastPartition() {
+    if (lastPartitionIndex == 0) {
+      return;
+    }
+
+    RowsAndColumns singleRac = MapOfColumnsRowsAndColumns.fromResultRowTillIndex(
+        rowsToProcess,
+        frameReader.signature(),
+        lastPartitionIndex
+    );
+    ArrayList<RowsAndColumns> rowsAndColumns = new ArrayList<>();
+    rowsAndColumns.add(singleRac);
+    runAllOpsOnMultipleRac(rowsAndColumns);
+
+    rowsToProcess.subList(0, lastPartitionIndex).clear();
+    lastPartitionIndex = 0;
+  }
 }
diff --git a/processing/src/main/java/org/apache/druid/query/rowsandcols/MapOfColumnsRowsAndColumns.java b/processing/src/main/java/org/apache/druid/query/rowsandcols/MapOfColumnsRowsAndColumns.java
index 42972f9340..3443166c0f 100644
--- a/processing/src/main/java/org/apache/druid/query/rowsandcols/MapOfColumnsRowsAndColumns.java
+++ b/processing/src/main/java/org/apache/druid/query/rowsandcols/MapOfColumnsRowsAndColumns.java
@@ -84,11 +84,16 @@ public class MapOfColumnsRowsAndColumns implements RowsAndColumns
   }
 
   public static MapOfColumnsRowsAndColumns fromResultRow(ArrayList<ResultRow> objs, RowSignature signature)
+  {
+    return fromResultRowTillIndex(objs, signature, objs.size());
+  }
+
+  public static MapOfColumnsRowsAndColumns fromResultRowTillIndex(ArrayList<ResultRow> objs, RowSignature signature, int index)
   {
     final Builder bob = builder();
     if (!objs.isEmpty()) {
-      Object[][] columnOriented = new Object[objs.get(0).length()][objs.size()];
-      for (int i = 0; i < objs.size(); ++i) {
+      Object[][] columnOriented = new Object[objs.get(0).length()][index];
+      for (int i = 0; i < index; ++i) {
         for (int j = 0; j < objs.get(i).length(); ++j) {
           columnOriented[j][i] = objs.get(i).get(j);
         }

```

All tests pass with the above diff. Let me know if this approach seems better to you compared to the current changes of the PR, I just need to update some comments before pushing this to the PR.

Reply:
@cryptoe @LakshSingla As discussed offline, I've pushed the above approach to the PR. Please have a look, thanks!

Also, I verified the behavior of `list.sublist(a,b).clear()`, it works as we expect it to (brings the later elements to the start, effectively resizes the list):
<img width="2276" alt="image" src="https://github.com/user-attachments/assets/08d6783a-5da9-4e13-aaf3-3ee65f3f1b67">